contracts/marketpools.js

```diff
@@ -0,0 +1,346 @@
+/* eslint-disable max-len */
+/* global actions, api */
+
+const TradeType = ['exactInput', 'exactOutput'];
+
+actions.createSSC = async () => {
+  const tableExists = await api.db.tableExists('pools');
+  if (tableExists === false) {
+    await api.db.createTable('pools', ['tokenPair']);
+    await api.db.createTable('liquidityPositions', ['account', 'tokenPair']);
+    await api.db.createTable('params');
+
+    const params = {};
+    params.poolCreationFee = '1000';
+    await api.db.insert('params', params);
+  }
+};
+
+actions.updateParams = async (payload) => {
+  const { poolCreationFee } = payload;
+  if (api.sender !== api.owner) return;
+  const params = await api.db.findOne('params', {});
+  if (poolCreationFee) {
+    if (!api.assert(typeof poolCreationFee === 'string' && !api.BigNumber(poolCreationFee).isNaN() && api.BigNumber(poolCreationFee).gte(0), 'invalid poolCreationFee')) return;
+    params.poolCreationFee = poolCreationFee;
+  }
+  await api.db.update('params', params);
+};
+
+function getAmountIn(amountOut, liquidityIn, liquidityOut) {
+  if (!api.assert(api.BigNumber(amountOut).gt(0), 'insufficient output amount')
+    || !api.assert(api.BigNumber(liquidityIn).gt(0)
+      && api.BigNumber(liquidityOut).gt(0)
+      && api.BigNumber(amountOut).lt(liquidityOut), 'insufficient liquidity')) return false;
+  const num = api.BigNumber(liquidityIn).times(amountOut);
+  const den = api.BigNumber(liquidityOut).minus(amountOut);
+  return num.dividedBy(den);
+}
+
+function getAmountOut(amountIn, liquidityIn, liquidityOut) {
+  if (!api.assert(api.BigNumber(amountIn).gt(0), 'insufficient output amount')
+    || !api.assert(api.BigNumber(liquidityIn).gt(0)
+      && api.BigNumber(liquidityOut).gt(0), 'insufficient liquidity')) return false;
+  const num = api.BigNumber(amountIn).times(liquidityOut);
+  const den = api.BigNumber(liquidityIn).plus(amountIn);
+  const amountOut = num.dividedBy(den);
+  if (!api.assert(api.BigNumber(amountOut).lt(liquidityOut), 'insufficient liquidity')) return false;
+  return amountOut;
+}
+
+async function validateOracle(pool, newPrice, maxDeviation = 0.01) {
+  const [baseSymbol, quoteSymbol] = pool.tokenPair.split(':');
+  // eslint-disable-next-line no-template-curly-in-string
+  const baseMetrics = baseSymbol !== "'${CONSTANTS.HIVE_PEGGED_SYMBOL}$'"
+    ? await api.db.findOneInTable('market', 'metrics', { symbol: baseSymbol })
+    : { lastPrice: 1 };
+  // eslint-disable-next-line no-template-curly-in-string
+  const quoteMetrics = quoteSymbol !== "'${CONSTANTS.HIVE_PEGGED_SYMBOL}$'"
+    ? await api.db.findOneInTable('market', 'metrics', { symbol: quoteSymbol })
+    : { lastPrice: 1 };
+  if (!baseMetrics || !quoteMetrics) return null; // no oracle available
+  const oracle = api.BigNumber(quoteMetrics.lastPrice).dividedBy(baseMetrics.lastPrice);
+  const dev = api.BigNumber(api.BigNumber(newPrice - oracle).abs()).dividedBy(oracle);
```

Comment:
newPrice - oracle should be changed to BigNumber subtraction

Reply:
Fixed